packages/caravan-health/src/waste.ts

```diff
@@ -0,0 +1,202 @@
+import { FeeRatePercentile, Transaction } from "@caravan/clients";
+import { utxoSetLengthScore } from "./privacy";
+import { AddressUtxos } from "./types";
+
+/*
+Utility function that helps to obtain the fee rate of the transaction
+
+Expected Range : [0, 0.75]
+-> Very Poor : [0, 0.15]
+-> Poor : (0.15, 0.3]
+-> Moderate : (0.3, 0.45]
+-> Good : (0.45, 0.6]
+-> Very Good : (0.6, 0.75]
+*/
+function getFeeRateForTransaction(transaction: Transaction): number {
+  let fees: number = transaction.fee;
+  let weight: number = transaction.weight;
+  return fees / weight;
+}
+
+/* 
+Utility function that helps to obtain the percentile of the fees paid by user in tx block
+
+Expected Range : [0, 0.75]
+-> 0% tile : 1
+-> 10% tile : 0.9
+-> 25% tile : 0.75
+-> 50% tile : 0.5
+-> 75% tile : 0.25
+-> 90% tile : 0.1
+-> 100% tile : 0.05
+*/
+function getFeeRatePercentileScore(
+  timestamp: number,
+  feeRate: number,
+  feeRatePercentileHistory: FeeRatePercentile[],
+): number {
+  let percentile: number = getPercentile(
+    timestamp,
+    feeRate,
+    feeRatePercentileHistory,
+  );
+  return 1-(percentile/100);
+}
+
+function getPercentile(
+  timestamp: number,
+  feeRate: number,
+  feeRatePercentileHistory: FeeRatePercentile[],
+): number {
+  // Find the closest entry by timestamp
+  let closestBlock: FeeRatePercentile | null = null;
+  let closestDifference: number = Infinity;
+
+  for (const block of feeRatePercentileHistory) {
+    const difference = Math.abs(block.timestamp - timestamp);
+    if (difference <= closestDifference) {
+      closestDifference = difference;
+      closestBlock = block;
+    }
+  }
```

Comment:
This is an example where block height would be more useful. You're not really looking for the closest block right? You're looking for the block the tx was mined in. If you have the block height for the tx then that's pretty straightforward as it's not a malleable value.

Reply:
Okay, I get it. But can you please confirm one think if that's the case.

https://mempool.space/api/v1/mining/blocks/fee-rates/all

In this API, We are getting something called avgHeight. So it's not the exact blockheight. But since the variable in incremental with time, I think it is related to block height.

So, what is more accurate in this case ? Looking at closest `avgHeight` or the closest `timestamp` ??

Reply:
Oh I see! Ok, I didn't fully understand this endpoint then. This makes sense to me. I thought it was the average fee rate for a block. I think since either or is basically a rough check, this is totally fine. We can always come back and address later if we also want block height to use for something else. Sorry for the confusion!